internal/proxy/oauthproxy.go

```diff
@@ -567,40 +565,29 @@ func (p *OAuthProxy) OAuthCallback(rw http.ResponseWriter, req *http.Request) {
 	}
 
 	// We validate the user information, and check that this user has proper authorization
-	// for the resources requested. This can be set via the email address or any groups.
+	// for the resources requested.
 	//
 	// set cookie, or deny
-	if !p.emailValidator(session.Email) {
-		tags = append(tags, "error:invalid_email")
+
+	errors := options.RunValidators(p.Validators, session)
+	if len(errors) != 0 {
+		tags = append(tags, "error:validation_failed")
 		p.StatsdClient.Incr("application_error", tags, 1.0)
 		logger.WithRemoteAddress(remoteAddr).WithUser(session.Email).Info(
-			"permission denied: unauthorized")
-		p.ErrorPage(rw, req, http.StatusForbidden, "Permission Denied", "Invalid Account")
-		return
-	}
-
-	allowedGroups := p.upstreamConfig.AllowedGroups
+			fmt.Sprintf("permission denied: unauthorized: %q", errors))
 
-	inGroups, validGroup, err := p.provider.ValidateGroup(session.Email, allowedGroups, session.AccessToken)
-	if err != nil {
-		tags = append(tags, "error:user_group_failed")
-		p.StatsdClient.Incr("provider_error", tags, 1.0)
-		logger.WithRemoteAddress(remoteAddr).WithUser(session.Email).Info(
-			"couldn't fetch user groups")
-		p.ErrorPage(rw, req, http.StatusInternalServerError, "Internal Error", "Error validating group membership, please try again")
-		return
-	}
-	if !validGroup {
-		tags = append(tags, "error:unauthorized_email")
-		p.StatsdClient.Incr("provider_error", tags, 1.0)
-		logger.WithRemoteAddress(remoteAddr).WithUser(session.Email).WithAllowedGroups(
-			allowedGroups).Info("permission denied: unauthorized")
-		p.ErrorPage(rw, req, http.StatusForbidden, "Permission Denied", "Group membership required")
+		formattedErrors := make([]string, 0, len(errors))
+		for _, err := range errors {
+			formattedErrors = append(formattedErrors, err.Error())
+		}
+		errorMsg := fmt.Sprintf("We ran into some issues while validating your account: \"%s\"",
+			strings.Join(formattedErrors, ", "))
+		p.ErrorPage(rw, req, http.StatusForbidden, "Permission Denied", errorMsg)
```

Comment:
TODO: Not a huge issue, but if any of the errors are duplicated (which is currently possible) then they'll also be duplicated on the error page -- which we could easily avoid

Reply:
SGTM. We could clean up the visual a little bit but let's do that in a follow up.